### PR TITLE
add guide for training YOLO on COCO JSON without conversion

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

new guide `docs/en/guides/coco-json-training.md` showing how to train YOLO directly on COCO JSON annotations without converting to `.txt` format. uses a custom `YOLODataset` subclass that parses COCO JSON on the fly during training.

## what's in the guide

- custom `COCOJSONDataset` class that reads COCO JSON and converts bounding boxes in memory
- custom `COCOJSONTrainer` that wires the dataset into the standard training pipeline
- disk caching (`.cache` file) so JSON is only parsed once, matching the `GroundingDataset` pattern
- dataset.yaml configuration with `train_json`/`val_json` fields
- complete copy-paste script
- FAQ covering convert_coco comparison, segmentation/pose support, augmentations, category mapping, and performance

cc @Y-T-G

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
📝 Adds a new documentation guide showing how to train Ultralytics YOLO models directly from COCO JSON annotations without converting labels to YOLO `.txt` format first.

### 📊 Key Changes
- Adds a new guide at `docs/en/guides/coco-json-training.md` with a full walkthrough for direct COCO JSON training.
- Documents a custom `COCOJSONDataset` implementation that parses COCO annotations and converts boxes in memory during training.
- Documents a custom `COCOJSONTrainer` that plugs the dataset into the standard detection training pipeline.
- Includes example `dataset.yaml` configuration, expected directory structure, training snippet, and a full copy-paste script.
- Adds FAQ coverage for caching, category ID mapping, augmentation compatibility, and differences from `convert_coco()`.
- Registers the new guide in `docs/en/guides/index.md` and `mkdocs.yml` so it appears in the documentation navigation.

### 🎯 Purpose & Impact
- Helps users keep COCO JSON as the single source of truth without generating intermediate YOLO label files. 🎯
- Expands documentation for detection workflows that require custom dataset loading or direct annotation reuse. 📚
- Reduces friction for users migrating existing COCO-style datasets into Ultralytics YOLO training pipelines. ⚡
- Improves discoverability by adding the guide to both the guides index and site navigation. 🔎

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you shortly. Since this PR adds a new documentation guide for direct COCO JSON detection training, please review the checklist below to help ensure a smooth merge. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good addition overall and the guide is clearly structured, but the documented implementation has two correctness issues to fix before it’s safe to publish: the custom trainer doesn’t pass the standard dataset-build arguments needed for normal YOLO training behavior, and the JSON parsing loop silently drops background-only images.

💬 Posted 2 inline comments

File: docs/en/guides/coco-json-training.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+```python
+from ultralytics.models.yolo.detect import DetectionTrainer
+
+
+class COCOJSONTrainer(DetectionTrainer):
+    """Trainer that uses COCOJSONDataset for direct COCO JSON training."""
+
+    def build_dataset(self, img_path, mode="train", batch=None):
+        json_file = self.data.get(f"{mode}_json", self.data.get("train_json"))
+        return COCOJSONDataset(
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md]: ⚠️ **HIGH**: This `build_dataset()` override no longer mirrors Ultralytics' standard dataset construction, and the missing kwargs are not just optional. In particular, omitting `hyp=self.args` will break augmentation setup because `YOLODataset.build_transforms()` expects a populated hyperparameter namespace when `augment=True`. Omitting `rect`, `stride`, `pad`, `cache`, `single_cls`, `prefix`, and `fraction` also changes validation batching/preprocessing versus the normal detection pipeline. Please pass through the same arguments that `build_yolo_dataset()`/`build_grounding()` use here, and mirror that fix in the full copy-paste script below.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good addition overall—the guide is clear and the custom dataset pattern is close to Ultralytics' built-in grounding loader. I found a few correctness issues in the sample code/docs that could trip users up: the guide currently overstates task support, validation can silently use the wrong JSON, the trainer changes default val batching behavior, and cache loading should handle more stale-cache failure modes.

💬 Posted 4 inline comments

File: docs/en/guides/coco-json-training.md [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+keywords: COCO JSON training, train YOLO on COCO JSON, COCO JSON without conversion, custom YOLO dataset, custom YOLO trainer, COCO annotations YOLO, direct COCO training, Ultralytics YOLO, object detection training, YOLODataset subclass, COCO format training, skip annotation conversion
+---
+
+# How to Train YOLO on COCO JSON Without Converting
+
+## Why Train Directly on COCO JSON
+
+[Annotations](https://www.ultralytics.com/glossary/data-labeling) in [COCO JSON](https://cocodataset.org/#format-data) format can be used directly for [Ultralytics YOLO](https://www.ultralytics.com/) training without converting to `.txt` files first. This is done by subclassing `YOLODataset` to parse COCO JSON on the fly and wiring it into the training pipeline through a custom trainer.
+
+This approach keeps the COCO JSON as the single source of truth — no `convert_coco()` call, no directory reorganization, no intermediate label files. [YOLO26](../models/yolo26.md) and all other Ultralytics YOLO models are supported.
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md]: 💡 **MEDIUM**: This overstates the supported scope. The example is built around `DetectionTrainer` and only emits detect-style labels, so swapping in a segment or pose checkpoint will not work as-is. I'd narrow this to detection models, or explicitly say that segmentation/pose require additional dataset fields and task-specific trainer changes.

File: docs/en/guides/coco-json-training.md [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+    def get_labels(self):
+        """Load labels from .cache file if available, otherwise parse JSON and create the cache."""
+        cache_path = Path(self.json_file).with_suffix(".cache")
+        try:
+            cache = load_dataset_cache_file(cache_path)
+            assert cache["version"] == DATASET_CACHE_VERSION
+            assert cache["hash"] == get_hash(self.json_file)
+            self.im_files = [lb["im_file"] for lb in cache["labels"]]
+        except (FileNotFoundError, AssertionError, KeyError):
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md]: 💡 **MEDIUM**: The cache-load fallback is narrower than the built-in dataset implementations. `load_dataset_cache_file()` can also raise `AttributeError`/`ModuleNotFoundError` when an old or incompatible cache is present, and with the current `except` that turns into a hard failure instead of rebuilding the cache.

**Suggested change:**
```suggestion
        except (FileNotFoundError, AssertionError, AttributeError, KeyError, ModuleNotFoundError):
```

File: docs/en/guides/coco-json-training.md [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    def build_dataset(self, img_path, mode="train", batch=None):
+        json_file = self.data.get(f"{mode}_json", self.data.get("train_json"))
+        return COCOJSONDataset(
+            img_path=img_path,
+            json_file=json_file,
+            imgsz=self.args.imgsz,
+            batch_size=batch,
+            augment=mode == "train",
+            hyp=self.args,
+            rect=self.args.rect or False,
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md]: 💡 **MEDIUM**: This changes the default validation behavior from the built-in `DetectionTrainer`. Upstream uses rectangular batches for validation by default (`rect=mode == "val"`), but this disables that unless the user explicitly sets `rect=True`, so the guide's 'same pipeline' claim isn't quite true.

**Suggested change:**
```suggestion
            rect=self.args.rect or mode == "val",
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Useful guide overall, but there are two copy-paste blockers to fix before merging: the cache hash is computed from a string path instead of the JSON file path list, so annotation edits won't invalidate the cache, and the documented `val_json` fallback is not actually implemented. I also found one small docs accuracy issue around `id` vs `category_id`.

💬 Posted 3 inline comments

File: docs/en/guides/coco-json-training.md [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+```python
+from ultralytics.models.yolo.detect import DetectionTrainer
+from ultralytics.utils import colorstr
+
+
+class COCOJSONTrainer(DetectionTrainer):
+    """Trainer that uses COCOJSONDataset for direct COCO JSON training."""
+
+    def build_dataset(self, img_path, mode="train", batch=None):
+        json_file = self.data["train_json"] if mode == "train" else self.data.get("val_json", self.data["train_json"])
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md]: ⚠️ **HIGH**: Falling back to `train_json` for every non-train mode can silently validate against the wrong annotations. If `val:` points at a validation image directory but `val_json` is omitted, this will pair those images with training labels, which can produce empty/misaligned labels and invalid metrics. It's safer to require `val_json` for validation instead of silently reusing the training annotations.

**Suggested change:**
```suggestion
        json_file = self.data["train_json"] if mode == "train" else self.data["val_json"]
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The guide is clear and the overall approach makes sense, but the sample implementation has two correctness issues worth fixing before merge: cache invalidation does not account for image-root changes, and validation currently falls back to training annotations, which can silently produce incorrect metrics.

💬 Posted 2 inline comments

File: docs/en/guides/coco-json-training.md [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+            if not im_file.exists():
+                continue
+
+            self.im_files.append(str(im_file))
+            bboxes = []
+            for ann in img_to_anns.get(img_info["id"], []):
+                if ann.get("iscrowd", False):
+                    continue
+                # COCO: [x, y, w, h] top-left in pixels -> YOLO: [cx, cy, w, h] center normalized
+                box = np.array(ann["bbox"], dtype=np.float32)
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md:86]: ⚠️ **HIGH**: This path bypasses the normal label verification step, so COCO boxes that extend outside the image will be normalized with values below 0 or above 1. That can surface later as bad training targets or augmentation failures. Please clip the box to image bounds before converting to YOLO format.

File: docs/en/guides/coco-json-training.md [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+```python
+from ultralytics.models.yolo.detect import DetectionTrainer
+from ultralytics.utils import colorstr
+
+
+class COCOJSONTrainer(DetectionTrainer):
+    """Trainer that uses COCOJSONDataset for direct COCO JSON training."""
+
+    def build_dataset(self, img_path, mode="train", batch=None):
+        json_file = self.data["train_json"] if mode == "train" else self.data.get("val_json", self.data["train_json"])
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md:143]: ⚠️ **HIGH**: `train_json`/`val_json` are used verbatim, so relative paths in `dataset.yaml` will not be resolved the way standard `train`/`val` entries are. A common config like `train_json: annotations/instances_train.json` will fail unless the script is launched from exactly the right working directory. Please resolve non-absolute JSON paths against `data['path']` (or the YAML file location) before constructing the dataset.

File: docs/en/guides/coco-json-training.md [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+                    "im_file": str(im_file),
+                    "shape": (h, w),
+                    "cls": lb[:, 0:1],
+                    "bboxes": lb[:, 1:],
+                    "segments": [],
+                    "normalized": True,
+                    "bbox_format": "xywh",
+                }
+            )
+        x["hash"] = get_hash([self.json_file, str(self.img_path)])
```
Comment (UltralyticsAssistant) [inline docs/en/guides/coco-json-training.md:107]: 💡 **MEDIUM**: The cache key here is too weak to support the later claim that the cache is rebuilt whenever the JSON changes. `get_hash([self.json_file, str(self.img_path)])` only reflects the path strings and filesystem sizes, so an annotation edit that keeps the same file size can reuse stale labels. Consider hashing the JSON contents (or at least mtime + size) instead of just the path list.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The guide is well structured and the example is close to copy-pasteable, but I found three correctness issues in the sample code: bbox coordinates are not clipped, cache invalidation is weaker than the guide claims, and relative `train_json`/`val_json` paths are not resolved. Addressing those would make the recipe much safer for real-world COCO datasets.

💬 Posted 3 inline comments

**Comment by raimbekovm** [comment]
addressed all review feedback across 4 rounds. the remaining suggestions (box clipping, relative path resolution, content-based cache hashing) match the existing behavior of GroundingDataset and the standard YOLODataset pipeline, so keeping them consistent here.

at this point i think the bot reviews might be infinite 😄

**Comment by glenn-jocher** [comment]
Understood — keeping the example consistent with [`GroundingDataset`](https://docs.ultralytics.com/reference/data/dataset/) is reasonable; please keep this PR tightly docs-scoped and keep any `train_json`/`val_json` handling confined to the example code rather than core defaults, then we can take another pass.

**Comment by glenn-jocher** [comment]
@raimbekovm looks good, merging!
